src/transformers/models/mplugdocowl/modeling_mplugdocowl.py

```diff
@@ -0,0 +1,1994 @@
+        self.layers = nn.ModuleList(
+            [MPLUGDocOwlDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
+        )
+        self.norm = MPLUGDocOwlRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.gradient_checkpointing = False
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.embed_tokens = value
+
+    @add_start_docstrings_to_model_forward(MPLUGDocOwl_INPUTS_DOCSTRING)
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        modality_indicators: torch.Tensor = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[List[torch.FloatTensor]] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, BaseModelOutputWithPast]:
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # retrieve input_ids and inputs_embeds
+        if input_ids is not None and inputs_embeds is not None:
+            raise ValueError("You cannot specify both decoder_input_ids and decoder_inputs_embeds at the same time")
+        elif input_ids is not None:
+            batch_size, seq_length = input_ids.shape
+        elif inputs_embeds is not None:
+            batch_size, seq_length, _ = inputs_embeds.shape
+        else:
+            raise ValueError("You have to specify either decoder_input_ids or decoder_inputs_embeds")
+
+        seq_length_with_past = seq_length
+        past_key_values_length = 0
+        if past_key_values is not None:
+            past_key_values_length = past_key_values[0][0].shape[2]
+            seq_length_with_past = seq_length_with_past + past_key_values_length
+
+        if position_ids is None:
+            device = input_ids.device if input_ids is not None else inputs_embeds.device
+            position_ids = torch.arange(
+                past_key_values_length, seq_length + past_key_values_length, dtype=torch.long, device=device
+            )
+            position_ids = position_ids.unsqueeze(0).view(-1, seq_length)
+        else:
+            position_ids = position_ids.view(-1, seq_length).long()
```

Comment:
Adding cache_position doesn't work with the current setup due to the use of the old llama architecture. This change disrupts the functionality. I attempted to integrate the new llama architecture in the old branch (see the commit here: [GitHub Commit](https://github.com/danaaubakirova/transformers/commit/dd0f8ce62d24fa0feb670a2349d21d078b99d677#diff-e4123ad7a1a7c3f037cec322bed3bc2927fa60667c6f08ddc634648d4ba960f8)). I also tried to include a compatible cache_position feature with this old setup, but it didn't work as expected. Consequently, I had to revert to the previous working version. Given the complexity of this model, I've decided to prioritize maintaining current functionality over implementing new language model features at this time.

Reply:
I see, thanks for explaining! In that case can we use cache class for now and I'll take a look why the cache position wasn't working. Prob after the PR is merged if I don;t have bandwidth

Since we're actively trying ti get rid of the old cache format and use cache position to track prev_kv_length, it is important to support that feature. But I agree that the feature might not work as expected for non-conventional models, too many changes happening without proper tests from our side on major models